enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,652 @@
+means to an end for implementing a OpenShift networking feature in the cloud (in
+this case: egress IP) by the networking stack. As such the primary "user" of
+this cloud component will be the network plugins themselves. This impacts the
+design decision with regards to the cloud component's failure notification, see
+[CRD section](#crd).
+
+##### network plugin <-> cloud API controller
+
+The network plugin needs to remain "source of truth" of the egress IP setup. The
+reason for this is that the network plugin is the one and only component which
+health checks routes to egress nodes and re-balances the egress IP assignment
+depending on the result. This component will need to execute what the network
+plugin has determined is best. This is done since there is no traffic load
+balancing done today for egress IP intended traffic. The network plugin
+constantly initiates a TCP connection to each egress node and depending on the
+success/failure of that: re-balances the egress assignment to other nodes (thus
+avoiding packets being sent to a egress node which is currently un-reachable).
+Going forward, both network plugins will come to support traffic load-balancing
+and a more sophisticated route health detection. That means that the underlying
+networking solution (OVS in the case of both OVN-Kubernetes and openshift-sdn)
+will manage the route health check. It is worth noting that the "amount of
+re-balancing" done by the network plugin is expected to be reduced, since the
+network fabric will be able to load balance traffic between all egress nodes
+provided and always use a healthy node (thus not requiring the network plugin to
+perform an egress IP re-assignment every time a route goes down).
+
+Given the above, it is however important that the network plugin is conveyed the
+cloud's information concerning the nodes' subnet range. This is especially
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes' default `/32` subnet.
+
+#### Inter-component communication
+
+Multiple pieces of information needs to be communicated between the network
+plugins and this cloud component:
+
+1. The cloud's node subnet and capacity information.
+2. The egress IP assignment to be performed by the cloud component
+3. The status of the egress IP assignment performed by the cloud component
+
+Two models have been decided for this: a CRD based approach and a node
+annotated approach.
+
+##### annotated approach
+
+As to satisfy condition 1. the cloud component will annotate all cluster nodes
+with the nodes' cloud subnet as retrieved from the cloud provider's API when it
+is initialized. This annotation will have the following specification:
+
+`cloud.openshift.io/private-ipconfig: {"ifaddr": {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}, "capacity": {"ipv4": "$IPv4_CAPACITY",  "ipv6": "$IPv6_CAPACITY"}}`
```

Comment:
The earlier text implied that capacity was address-family-agnostic... though it seems like it could easily be the case that there are IP limits per family on some clouds and IP limits across all families on other clouds. We probably need to be able to explain it both ways...

Also, is it guaranteed that all clouds will have at most one IPv4 subnet and at most one IPv6 subnet per interface? And we're probably going to eventually support multiple interfaces anyway. I'd do something more like:

    [{"interface": "$IFNAME", capacity: $CAPACITY, subnets: [ "$CIDR1", "$CIDR2", ... ]}]

(You don't need to explicitly indicate IPv4 vs IPv6 CIDRs; you can tell when you parse them.)

or maybe just change the text to be vaguer ("This annotation will look something like the following:")

Reply:
> The earlier text implied that capacity was address-family-agnostic... though it seems like it could easily be the case that there are IP limits per family on some clouds and IP limits across all families on other clouds. 

OK, I will make sure to revise the text. Capacity needs to be per IP family, because on AWS the values differ based on instance type and IP family. 

> is it guaranteed that all clouds will have at most one IPv4 subnet **and** at most one IPv6 subnet per interface?

No, it's an **or**. I think it's guaranteed that a interface will have on or the other, possibly both. At least in the single-stack case. It might be "and" in the dual-stack case, but I have not created a dual-stack cluster on GCP / AWS / Azure and verified...can I even do that?

> I'd do something more like:

That won't work for AWS if the node you're assigning an egress IP to is of, for example, instance type: c1.medium. See: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#AvailableIpPerENI we really need capacity to be defined per IP family as to ensure we cover AWS correctly.  Moreover, `$IFNAME` won't mean anything to the network plugins (or to anyone at all, for that matter), the clouds will just use a cryptic ID they've defined for the interface and we won't be able to tell it a part. Maybe `$IFIDX` makes more sense if we're going for something like this. 

> or maybe just change the text to be vaguer ("This annotation will look something like the following:")

Yeah, maybe this is a topic of discussion better addressed on the PR implementing the functionlity.

Reply:
> Capacity needs to be per IP family, because on AWS the values differ based on instance type and IP family.

Right, but couldn't there be other clouds that _don't_ do it that way? eg, maybe Alibaba Cloud has a max of 50 total additional IPs, combined between IPv4 and IPv6. In that case if the annotation forces you to specify separate IPv4 and IPv6 capacities, you'd have to just arbitrarily divide them up and claim it allows 25 v4 and 25 v6 or something like that, even though the user might only want to use one family.

So I think the annotation has to be able to express _either_ "you can have N IPv4 IPs and M IPv6 IPs" _or_ "you can have a total of N IPv4+IPv6 IPs", and then the network plugin has to be able to work with both types.

> It might be "and" in the dual-stack case, but I have not created a dual-stack cluster on GCP / AWS / Azure and verified...can I even do that?

At the moment we don't support dual stack on any clouds, but we probably will in the future so the API should be ready for that. (And in fact, none of the cloud types we currently support allows single-stack IPv6; only dual-stack.)

> Moreover, `$IFNAME` won't mean anything to the network plugins (or to anyone at all, for that matter), the clouds will just use a cryptic ID they've defined for the interface and we won't be able to tell it a part.

We have access to the API, we ought to be able to figure it out. Eg, you ought to be able to find the MAC address that goes with the cryptic ID and compare that to the interfaces on the node.

Reply:
> So I think the annotation has to be able to express either "you can have N IPv4 IPs and M IPv6 IPs" or "you can have a total of N IPv4+IPv6 IPs", and then the network plugin has to be able to work with both types.

Agreed. Maybe something like this fits:

on clouds where the IPv4 vs IPv6 capacity distinction exists, we can use:

```
`cloud.network.openshift.io/egress-ipconfig: {"ifaddr": {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}, "capacity": {"ipv4": "$IPv4_CAPACITY",  "ipv6": "$IPv6_CAPACITY"}}`
```

on clouds where that is not the case:

```
`cloud.network.openshift.io/egress-ipconfig: {"ifaddr": {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}, "capacity": {"ip": "$IPv4_AND_IPv6_CAPACITY"}}`
```

applied with an XOR between both cases.

> We have access to the API, we ought to be able to figure it out.

De-coupling the network plugins from the cloud (and credential management, etc) was one of the main reasons for using a dedicated component which should do that for them. I don't think we should have the network plugins talk to the cloud on their end and figure out what the NIC ID / name corresponds to by themselves....if that's what you were pointing to? 

> Eg, you ought to be able to find the MAC address that goes with the cryptic ID and compare that to the interfaces on the node

Not really, a "NIC" on GCP does not even have a MAC address (at least a user-exposed MAC address) :shrug: 

In any case: I think we're really in a situation where we're at step 1. and thinking about a possible solution for step 3. without ever having spent some thought on step 2. (which in this case would be: multi-NIC support for egress IP on bare-metal). 

I think we can do what you mentioned in the beginning for now, something along the lines of:

```
`cloud.network.openshift.io/egress-ipconfig: [{"interface": "$IFNAME/$IFID", "ifaddr": {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}, "capacity": {"ipv4": "$IPv4_CAPACITY",  "ipv6": "$IPv6_CAPACITY"}}]`
```

Which for now would be a list of length 1 for now, which only does what it needs to for the primary NIC. That will, at the very least, lead to less change for whatever we might do in the future.